File: catalyst-backend/functions/maptix_api/routes/export.js
'use strict';

/**
 * Export Routes — Download models in various formats
 * 
 * GET /v1/projects/:projectId/export?format=svg  — Export as SVG
 * GET /v1/projects/:projectId/export?format=json — Export as JSON
 */

const express = require('express');
const router = express.Router({ mergeParams: true });
const { requireAuth } = require('../lib/auth-middleware');
const { getSpatialModel } = require('../lib/datastore');
const { generateSVG } = require('../lib/svg-exporter');

router.use(requireAuth);

router.get('/', async (req, res, next) => {
  try {
    const { projectId } = req.params;
    const format = req.query.format;

    if (!format) {
      return res.status(400).json({ detail: 'format query parameter is required (svg or json)' });
    }

    if (!req.catalystApp) {
      return res.status(404).json({ detail: 'No model found to export' });
    }

    const model = await getSpatialModel(req.catalystApp, projectId);
    if (!model) {
      return res.status(404).json({ detail: 'No model found to export' });
    }

    const modelData = model.model_data || (model.model_data_json ? JSON.parse(model.model_data_json) : {});

    if (format === 'svg') {
      const svgContent = generateSVG(modelData);
      res.set({
        'Content-Type': 'image/svg+xml',
        'Content-Disposition': 'attachment; filename="floorplan.svg"',
      });
      return res.send(svgContent);
    }

    if (format === 'json') {
      const jsonContent = JSON.stringify(modelData, null, 2);
      res.set({
        'Content-Type': 'application/json',
        'Content-Disposition': 'attachment; filename="spatial-model.json"',
      });
      return res.send(jsonContent);
    }

    res.status(400).json({
      detail: `Export format '${format}' not supported. Available: svg, json`,
    });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: backend/app/api/v1/endpoints/export.py
"""
Export endpoints — download models in various formats.
"""

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import joinedload

from app.db.session import get_db
from app.db.models import Project, SpatialModel
from app.core.auth import get_current_user

from app.services.export.svg_exporter import generate_svg

router = APIRouter()


@router.get("")
async def export_model(
    project_id: str,
    format: str = Query(...),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    """Export the project model in the specified format."""
    # Verify project ownership
    project_q = select(Project).where(
        Project.id == project_id, Project.owner_id == current_user.id
    )
    project = (await db.execute(project_q)).scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    query = select(SpatialModel).where(SpatialModel.project_id == project_id)
    result = await db.execute(query)
    model = result.scalar_one_or_none()

    if not model:
        raise HTTPException(status_code=404, detail="No model found to export")

    if format == "svg":
        svg_content = generate_svg(model.model_data)
        return StreamingResponse(
            iter([svg_content.encode()]),
            media_type="image/svg+xml",
            headers={"Content-Disposition": 'attachment; filename="floorplan.svg"'},
        )

    if format == "json":
        import json
        json_content = json.dumps(model.model_data, indent=2)
        return StreamingResponse(
            iter([json_content.encode()]),
            media_type="application/json",
            headers={"Content-Disposition": 'attachment; filename="spatial-model.json"'},
        )

    raise HTTPException(
        status_code=400,
        detail=f"Export format '{format}' not supported. Available: svg, json"
    )
